### PR TITLE
Run E2E integration tests on pull requests and branch build

### LINKED ISSUE #615: Prepare integration tests based on local Kubernetes cluster
## Description

- Prepare basic test scenario with Slack
    - Check if events are posted on channel, and if BotKube reacts for some kubectl commands 
    - Ideally use real Slack with channel create/delete; then these tests will be run on `develop` branch
    - Alternatively, use Slack mock deployed in cluster; then we will be able to run these tests on pull requests 
- Use real K8s API, executor and deploy BotKube with Helm
- Spawn k3d/kind cluster and install Helm chart
- Run tests via `helm test`
- Remove duplicated test cases from current tests based on fake K8s API
- Prepare GitHub pipeline (if we use mocked Slack, then it is both for PR and `develop` and it depends on #590; if not, then it runs only on `develop`)

## Related issue

See the epic #589

### PR DESCRIPTION
**Pull request by pkosiec**
##### ISSUE TYPE
 - Feature Pull Request

##### SUMMARY

~⚠️ This PR depends on #622, which should be merged first. Then I rebase this one 🙂 
Actual changes: https://github.com/pkosiec/botkube/compare/integration-tests...pkosiec:integration-tests-ci?expand=1~

This PR introduces PR tests only. I believe we should improve branch build pipeline soon (e.g. to have reproducible tests, we should push Docker images with commit hash tag, same with Helm chart version). I can prepare a separate task for that.
Alternatively, after merging this, I can do a follow-up PR to introduce integration tests step in the `artifacts.yaml` build (I can rename it to `branch-build.yaml`) to have something in the meantime. But then the tests would base on `v9.99.9-dev` image and Helm chart.

- Run E2E integration tests on branch build
- Use Goreleaser to build test image

Resolves #615

##### TESTING

You can follow README.md in `tests` directory to run tests via Helm.

Tested on my fork: https://github.com/pkosiec/botkube/runs/7018919907?check_suite_focus=true (PR: https://github.com/pkosiec/botkube/pull/1)

Screenshot 2022-06-23 at 09 41 52

### CODE REVIEW

File: helm/botkube/values.yaml [hunk 1]
```diff
@@ -505,3 +505,18 @@ extraObjects: []
 #      - kind: ServiceAccount
 #        name: "{{ include \"botkube.serviceAccountName\" . }}"
 #        namespace: "{{ .Release.Namespace }}"
+
+## Parameters for the test container with E2E tests.
+e2eTest:
+  image:
+    registry: docker.io
+    repository: pkosiec/botkube-test # TODO: Replace with image built on `develop`
+    tag: latest
```
Comment (pkosiec) [inline helm/botkube/values.yaml:514]: I will replace this in a follow-up PR after merging this PR and building test image.

File: tests/README.md [hunk 2]
```diff
@@ -41,57 +41,47 @@ This directory contains E2E tests which are run against BotKube installed on Kub
...
+    export IMAGE_REPOSITORY="infracloudio/botkube"
+    export IMAGE_TAG="v9.99.9-dev"
+
+    #
+    # The following environmental variables are required only when running integration tests via Helm:
+    #
+    export SLACK_TESTER_APP_TOKEN="{BotKube tester app token}" # WARNING: This is a token for Tester, not the BotKube Slack bot!
+    export TEST_IMAGE_REGISTRY="docker.io" # TODO: Replace with image built on `develop`
+    export TEST_IMAGE_REPOSITORY="pkosiec/botkube-test" # TODO: Replace with image built on `develop`
+    export TEST_IMAGE_TAG="latest" # TODO: Replace with image built on `develop`
```
Comment (pkosiec) [inline tests/README.md:66]: I will replace this in a follow-up PR after merging this PR and building test image.

File: .github/workflows/pr-build.yaml [hunk 3]
```diff
@@ -1,14 +1,13 @@
-name: Publish BotKube image
+name: Pull request Build
 
 on:
   pull_request_target:
     types: [ opened, synchronize, reopened ]
     branches:
-      - "develop"
+      - "main"
```
[the remark this answers is not in the document]
Comment (pkosiec) [reply .github/workflows/pr-build.yaml:7]: Oops, sorry, that was a leftover from my testing. Will fix that 👍

File: .github/workflows/pr-build.yaml [hunk 4]
```diff
@@ -114,8 +121,57 @@ jobs:
...
-              --set image.tag=${PR_NUMBER}-PR \\
+              --set image.repository=${IMAGE_REPOSITORY} \\
+              --set image.tag=${IMAGE_TAG} \\
+              --set e2eTest.image.repository=${TEST_IMAGE_REPOSITORY} \\
+              --set e2eTest.image.tag=${IMAGE_TAG} \\
               ./helm/botkube
 
           ENDOFFILE
+
+  integration-tests:
```
[the remark this answers is not in the document]
Comment (pkosiec) [reply .github/workflows/pr-build.yaml:132]: Yep, will do that on the same PR 👍 Nice idea with reusing workflow, however as you don't mind, I will do copy-paste. We can later improve that as you wrote - maybe when envs are accessible in the reusable workflows 🤔

File: helm/botkube/values.yaml (hunk 1 above)
Comment (pkosiec) [reply helm/botkube/values.yaml:514]: Yes, you're totally right 👍 I will change that before merge then

**Comment by pkosiec** [comment]
Thanks @mszostok - tested:
Branch build: https://github.com/drill-bot/botkube/runs/7129446709?check_suite_focus=true
PR build: https://github.com/drill-bot/botkube/pull/2

**Review by mszostok — APPROVED** [review]

File: hack/release.sh [hunk 5]
```diff
@@ -48,19 +48,21 @@ generate_changelog() {
...
 }
 
-update_chart_yamls $version
-generate_changelog $version
-make release
+fetch_previous_version
+update_chart_yamls $previous_version $version
+update_chart_yamls "v9.99.9-dev" $version
+#generate_changelog $version
+#make release
```
[the remark this answers is not in the document]
Comment (pkosiec) [reply hack/release.sh:65]: The lines shouldn't be commented. Thanks for pointing that out 👍

**Pull request merged**